File: backend/src/application/pool_operation_service.py
"""池操作应用服务"""

from __future__ import annotations

import logging
from typing import Optional

from sqlalchemy.ext.asyncio import AsyncSession

from application.dto.pool_operation import (
    PoolKlineCollectRequest,
    PoolOperationListRequest,
    PoolOperationVO,
    PoolOperationListResponse,
    PoolOperationCreateResponse,
    PoolOperationProgressVO,
)
from application.exceptions import (
    PoolNotFoundError,
    PoolOperationNotFoundError,
    PoolOperationConflictError,
)
from domain.stock_pool.repository import StockPoolRepository, PoolOperationRepository
from infrastructure.repositories.pool_repository import StockPoolRepoImpl
from infrastructure.repositories.pool_operation_repository import PoolOperationRepoImpl
from infrastructure.tasks.operation_dispatcher import OperationDispatcher

logger = logging.getLogger(__name__)


class PoolOperationAppService:
    """池操作应用服务"""

    def __init__(self, session: AsyncSession):
        self._session = session
        self._pool_repo: StockPoolRepository = StockPoolRepoImpl(session)
        self._op_repo: PoolOperationRepository = PoolOperationRepoImpl(session)
        self._dispatcher = OperationDispatcher(session)

    async def create_kline_collect_operation(
        self, request: PoolKlineCollectRequest
    ) -> PoolOperationCreateResponse:
        pool = await self._pool_repo.find_by_id(request.pool_id)
        if not pool:
            raise PoolNotFoundError(request.pool_id)

        existing = await self._op_repo.list_by_pool(request.pool_id, limit=1)
        for op in existing:
            if op["status"] in ("pending", "running"):
                raise PoolOperationConflictError(request.pool_id)

        members = await self._pool_repo.list_members(request.pool_id, limit=10000)
        symbols = [m.symbol for m in members]

        if not symbols:
            return PoolOperationCreateResponse(
                operation_id=0,
                pool_id=request.pool_id,
                operation_type=request.operation_type,
                status="skipped",
                total=0,
                message="池内没有成员",
            )

        params = {
            "days": request.days,
            "source": request.source,
        }
        op_id = await self._op_repo.create(
            pool_id=request.pool_id,
            operation_type=request.operation_type,
            params=params,
        )

        await self._op_repo.update_progress(
            op_id, done=0, total=len(symbols)
        )

        await self._dispatcher.dispatch_kline_collect(
            op_id=op_id,
            pool_id=request.pool_id,
            symbols=symbols,
            days=request.days,
            source=request.source,
        )

        return PoolOperationCreateResponse(
            operation_id=op_id,
            pool_id=request.pool_id,
            operation_type=request.operation_type,
            status="pending",
            total=len(symbols),
            message=f"操作已派发，共 {len(symbols)} 只股票",
        )

    async def list_operations(
        self, request: PoolOperationListRequest
    ) -> PoolOperationListResponse:
        pool = await self._pool_repo.find_by_id(request.pool_id)
        if not pool:
            raise PoolNotFoundError(request.pool_id)

        ops = await self._op_repo.list_by_pool(
            request.pool_id, limit=request.limit, offset=request.offset
        )

        items = []
        for op in ops:
            progress = PoolOperationProgressVO(
                done=op["progress"].get("done", 0),
                total=op["progress"].get("total", 0),
                failed=op["progress"].get("failed", 0),
            )
            items.append(
                PoolOperationVO(
                    id=op["id"],
                    pool_id=op["pool_id"],
                    operation_type=op["operation_type"],
                    status=op["status"],
                    params=op["params"],
                    result_summary=op["result_summary"],
                    progress=progress,
                    error_message=op["error_message"],
                    started_at=op["started_at"],
                    finished_at=op["finished_at"],
                    created_at=op["created_at"],
                )
            )

        return PoolOperationListResponse(
            pool_id=request.pool_id,
            total=len(items),
            items=items,
        )

    async def get_operation(self, op_id: int) -> PoolOperationVO:
        op = await self._op_repo.find_by_id(op_id)
        if not op:
            raise PoolOperationNotFoundError(op_id)

        progress = PoolOperationProgressVO(
            done=op["progress"].get("done", 0),
            total=op["progress"].get("total", 0),
            failed=op["progress"].get("failed", 0),
        )
        return PoolOperationVO(
            id=op["id"],
            pool_id=op["pool_id"],
            operation_type=op["operation_type"],
            status=op["status"],
            params=op["params"],
            result_summary=op["result_summary"],
            progress=progress,
            error_message=op["error_message"],
            started_at=op["started_at"],
            finished_at=op["finished_at"],
            created_at=op["created_at"],
        )

    async def get_operation_progress(
        self, op_id: int
    ) -> PoolOperationProgressVO:
        op = await self._op_repo.find_by_id(op_id)
        if not op:
            raise PoolOperationNotFoundError(op_id)

        return PoolOperationProgressVO(
            done=op["progress"].get("done", 0),
            total=op["progress"].get("total", 0),
            failed=op["progress"].get("failed", 0),
        )

    async def cancel_operation(self, op_id: int) -> PoolOperationVO:
        op = await self._op_repo.find_by_id(op_id)
        if not op:
            raise PoolOperationNotFoundError(op_id)

        if op["status"] not in ("pending", "running"):
            from application.exceptions import ApplicationError
            raise ApplicationError("该操作无法取消", "CANNOT_CANCEL")

        await self._dispatcher.cancel_operation(op_id)
        await self._op_repo.update_status(
            op_id,
            status="cancelled",
            finished_at=op.get("started_at"),
        )
        return await self.get_operation(op_id)

File: backend-java/src/main/java/com/attribution/application/service/PoolOperationService.java
package com.attribution.application.service;

import com.attribution.domain.entity.PoolOperationEntity;
import com.attribution.domain.entity.StockPoolEntity;
import com.attribution.interfaces.dto.operation.PoolKlineCollectRequest;
import com.attribution.interfaces.dto.operation.PoolOperationCreateVO;
import com.attribution.interfaces.dto.operation.PoolOperationListRequest;
import com.attribution.interfaces.dto.operation.PoolOperationListVO;
import com.attribution.interfaces.dto.operation.PoolOperationProgressVO;
import com.attribution.interfaces.dto.operation.PoolOperationVO;
import com.attribution.infrastructure.exception.BusinessException;
import org.springframework.http.HttpStatus;
import com.attribution.domain.repository.PoolOperationRepository;
import com.attribution.domain.repository.StockPoolMemberRepository;
import com.attribution.domain.repository.StockPoolRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class PoolOperationService {

    private final StockPoolRepository poolRepository;
    private final StockPoolMemberRepository memberRepository;
    private final PoolOperationRepository operationRepository;
    private final OperationDispatcher operationDispatcher;

    @Transactional
    public PoolOperationCreateVO createKlineCollectOperation(PoolKlineCollectRequest request) {
        StockPoolEntity pool = poolRepository.findById(request.getPoolId())
            .orElseThrow(() -> new BusinessException("pool", "POOL_NOT_FOUND", "操作池不存在: " + request.getPoolId(), HttpStatus.NOT_FOUND));

        List<PoolOperationEntity> existing = operationRepository.findByStatusIn(List.of("pending", "running"));
        boolean conflict = existing.stream()
            .anyMatch(op -> op.getPoolId() != null && op.getPoolId().equals(request.getPoolId()));
        if (conflict) {
            throw new BusinessException("operation", "OPERATION_CONFLICT", "_" + request.getPoolId() + " 存在未完成的操作", HttpStatus.CONFLICT);
        }

        List<String> symbols = memberRepository.findByPoolIdOrderBySortOrderAsc(request.getPoolId())
            .stream()
            .map(m -> m.getSymbol())
            .toList();

        if (symbols.isEmpty()) {
            return PoolOperationCreateVO.builder()
                .operationId(0L)
                .poolId(request.getPoolId())
                .operationType(request.getOperationType())
                .status("skipped")
                .total(0)
                .message("池内没有成员")
                .build();
        }

        Map<String, Object> params = new HashMap<>();
        params.put("days", request.getDays());

        PoolOperationEntity entity = PoolOperationEntity.builder()
            .poolId(request.getPoolId())
            .operationType(request.getOperationType())
            .status("pending")
            .params(params)
            .progress(Map.of("done", 0, "total", symbols.size(), "failed", 0))
            .build();
        PoolOperationEntity saved = operationRepository.save(entity);

        operationDispatcher.dispatchKlineCollect(saved.getId(), request.getPoolId(), symbols, request.getDays());

        return PoolOperationCreateVO.builder()
            .operationId(saved.getId())
            .poolId(request.getPoolId())
            .operationType(request.getOperationType())
            .status("pending")
            .total(symbols.size())
            .message("操作已派发，_" + symbols.size() + " 只股")
            .build();
    }

    @Transactional(readOnly = true)
    public PoolOperationListVO listOperations(PoolOperationListRequest request) {
        if (!poolRepository.existsById(request.getPoolId())) {
            throw new BusinessException("pool", "POOL_NOT_FOUND", "操作池不存在: " + request.getPoolId(), HttpStatus.NOT_FOUND);
        }
        List<PoolOperationEntity> ops = operationRepository.findByPoolIdOrderByCreatedAtDesc(request.getPoolId());
        List<PoolOperationVO> items = ops.stream()
            .skip(request.getOffset() != null ? request.getOffset() : 0)
            .limit(request.getLimit() != null ? request.getLimit() : ops.size())
            .map(PoolOperationVO::fromEntity)
            .toList();
        return PoolOperationListVO.builder()
            .poolId(request.getPoolId())
            .total(items.size())
            .dataList(items)
            .build();
    }

    @Transactional(readOnly = true)
    public PoolOperationVO getOperation(Long opId) {
        PoolOperationEntity entity = operationRepository.findById(opId)
            .orElseThrow(() -> new BusinessException("operation", "OPERATION_NOT_FOUND", "操作不存_ " + opId, HttpStatus.NOT_FOUND));
        return PoolOperationVO.fromEntity(entity);
    }

    @Transactional(readOnly = true)
    public PoolOperationProgressVO getOperationProgress(Long opId) {
        PoolOperationEntity entity = operationRepository.findById(opId)
            .orElseThrow(() -> new BusinessException("operation", "OPERATION_NOT_FOUND", "操作不存_ " + opId, HttpStatus.NOT_FOUND));
        return PoolOperationVO.fromEntity(entity).getProgress();
    }

    @Transactional
    public PoolOperationVO cancelOperation(Long opId) {
        PoolOperationEntity entity = operationRepository.findById(opId)
            .orElseThrow(() -> new BusinessException("operation", "OPERATION_NOT_FOUND", "操作不存_ " + opId, HttpStatus.NOT_FOUND));
        if (!List.of("pending", "running").contains(entity.getStatus())) {
            throw new BusinessException("operation", "CANNOT_CANCEL", "该操作无法取", HttpStatus.BAD_REQUEST);
        }
        operationDispatcher.cancel(opId);
        entity.setStatus("cancelled");
        PoolOperationEntity saved = operationRepository.save(entity);
        return PoolOperationVO.fromEntity(saved);
    }
}
